File: ios/MarrakechGuide/MarrakechGuide/Core/Services/LocationService.swift
import Foundation
import CoreLocation
import os.log

/// Protocol for location services providing GPS coordinates.
///
/// This service wraps CLLocationManager with battery-safe patterns:
/// - Request only "When In Use" location (foreground only)
/// - Start/stop updates based on screen visibility
/// - Automatic timeout to prevent battery drain
/// - Graceful handling of permission denial
///
/// Usage:
/// ```swift
/// let locationService = LocationServiceImpl.shared
/// await locationService.requestPermission()
/// locationService.startUpdates()
/// // ... use currentLocation
/// locationService.stopUpdates()
/// ```
public protocol LocationService: AnyObject, Sendable {
    /// Current user location, nil if unknown or permission denied.
    var currentLocation: CLLocation? { get }

    /// Current authorization status for location services.
    var locationAuthorizationStatus: CLAuthorizationStatus { get }

    /// Whether location services are authorized and available.
    var isAuthorized: Bool { get }

    /// Requests location permission from the user.
    /// - Returns: true if permission was granted, false otherwise.
    @MainActor
    func requestPermission() async -> Bool

    /// Starts receiving location updates.
    /// Updates are throttled to preserve battery.
    /// Call stopUpdates() when compass screen is no longer visible.
    func startUpdates()

    /// Stops receiving location updates.
    /// Always call this when the compass screen closes.
    func stopUpdates()

    /// Requests a single high-accuracy location refresh.
    /// Use sparingly as this is more battery-intensive.
    /// - Returns: The refreshed location.
    /// - Throws: LocationError if location cannot be obtained.
    func refreshLocation() async throws -> CLLocation
}

/// Errors that can occur during location operations.
public enum LocationError: Error, LocalizedError {
    case permissionDenied
    case locationUnavailable
    case timeout
    case servicesDisabled

    public var errorDescription: String? {
        switch self {
        case .permissionDenied:
            return "Location permission was denied"
        case .locationUnavailable:
            return "Unable to determine your location"
        case .timeout:
            return "Location request timed out"
        case .servicesDisabled:
            return "Location services are disabled"
        }
    }
}

/// Implementation of LocationService using CLLocationManager.
///
/// Thread-safety: All CLLocationManager operations happen on the main thread.
/// Public interface is Sendable for use from async contexts.
public final class LocationServiceImpl: NSObject, LocationService, @unchecked Sendable {

    // MARK: - Singleton

    public static let shared = LocationServiceImpl()

    // MARK: - Private Properties

    private let locationManager: CLLocationManager
    private let logger = Logger(subsystem: "com.marrakechguide", category: "LocationService")

    /// Lock for thread-safe access to mutable state
    private let lock = NSLock()

    /// Stored current location (accessed with lock)
    private var _currentLocation: CLLocation?

    /// Whether updates are currently active
    private var isUpdating = false

    /// Timeout timer for automatic stop
    private var timeoutTimer: Timer?

    /// Timeout duration in seconds (10 minutes default)
    private let timeoutDuration: TimeInterval = 600

    /// Continuation for permission request
    private var permissionContinuation: CheckedContinuation<Bool, Never>?

    /// Continuation for refresh request
    private var refreshContinuation: CheckedContinuation<CLLocation, Error>?

    /// Flag indicating we're waiting for a high-accuracy refresh
    private var isRefreshing = false

    // MARK: - Public Properties

    public var currentLocation: CLLocation? {
        lock.lock()
        defer { lock.unlock() }
        return _currentLocation
    }

    public var locationAuthorizationStatus: CLAuthorizationStatus {
        locationManager.authorizationStatus
    }

    public var isAuthorized: Bool {
        let status = locationAuthorizationStatus
        return status == .authorizedWhenInUse || status == .authorizedAlways
    }

    // MARK: - Initialization

    private override init() {
        locationManager = CLLocationManager()
        super.init()

        locationManager.delegate = self
        // Use balanced accuracy by default (battery-friendly)
        locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters
        // Distance filter to reduce update frequency
        locationManager.distanceFilter = 5.0
    }

    // MARK: - Public Methods

    @MainActor
    public func requestPermission() async -> Bool {
        let currentStatus = locationAuthorizationStatus

        switch currentStatus {
        case .notDetermined:
            // Request permission and wait for response
            return await withCheckedContinuation { continuation in
                self.permissionContinuation = continuation
                self.locationManager.requestWhenInUseAuthorization()
            }

        case .authorizedWhenInUse, .authorizedAlways:
            return true

        case .denied, .restricted:
            logger.info("Location permission denied or restricted")
            return false

        @unknown default:
            logger.warning("Unknown authorization status: \(String(describing: currentStatus))")
            return false
        }
    }

    public func startUpdates() {
        guard isAuthorized else {
            logger.info("Cannot start updates: not authorized")
            return
        }

        guard CLLocationManager.locationServicesEnabled() else {
            logger.info("Cannot start updates: location services disabled")
            return
        }

        guard !isUpdating else {
            logger.debug("Updates already active")
            return
        }

        isUpdating = true
        locationManager.startUpdatingLocation()
        logger.info("Started location updates")

        // Set timeout to prevent indefinite battery drain
        resetTimeoutTimer()
    }

    public func stopUpdates() {
        guard isUpdating else { return }

        isUpdating = false
        locationManager.stopUpdatingLocation()
        cancelTimeoutTimer()
        logger.info("Stopped location updates")
    }

    public func refreshLocation() async throws -> CLLocation {
        guard isAuthorized else {
            throw LocationError.permissionDenied
        }

        guard CLLocationManager.locationServicesEnabled() else {
            throw LocationError.servicesDisabled
        }

        // If we already have a recent location, return it
        if let location = currentLocation,
           Date().timeIntervalSince(location.timestamp) < 5 {
            return location
        }

        // Request high-accuracy location
        return try await withCheckedThrowingContinuation { continuation in
            self.refreshContinuation = continuation
            self.isRefreshing = true

            // Temporarily increase accuracy
            self.locationManager.desiredAccuracy = kCLLocationAccuracyBest

            if !self.isUpdating {
                self.locationManager.startUpdatingLocation()
            }

            // Timeout after 15 seconds
            DispatchQueue.main.asyncAfter(deadline: .now() + 15) { [weak self] in
                guard let self = self, self.isRefreshing else { return }
                self.isRefreshing = false
                self.refreshContinuation?.resume(throwing: LocationError.timeout)
                self.refreshContinuation = nil

                // Restore balanced accuracy
                self.locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters

                if !self.isUpdating {
                    self.locationManager.stopUpdatingLocation()
                }
            }
        }
    }

    // MARK: - Private Methods

    private func resetTimeoutTimer() {
        cancelTimeoutTimer()

        timeoutTimer = Timer.scheduledTimer(withTimeInterval: timeoutDuration, repeats: false) { [weak self] _ in
            self?.logger.warning("Location updates timed out after \(self?.timeoutDuration ?? 0) seconds")
            self?.stopUpdates()
        }
    }

    private func cancelTimeoutTimer() {
        timeoutTimer?.invalidate()
        timeoutTimer = nil
    }

    private func updateLocation(_ location: CLLocation) {
        lock.lock()
        _currentLocation = location
        lock.unlock()
    }
}

// MARK: - CLLocationManagerDelegate

extension LocationServiceImpl: CLLocationManagerDelegate {

    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        let status = manager.authorizationStatus
        logger.info("Authorization status changed: \(String(describing: status))")

        // Resume permission continuation if waiting
        if let continuation = permissionContinuation {
            permissionContinuation = nil
            let granted = status == .authorizedWhenInUse || status == .authorizedAlways
            continuation.resume(returning: granted)
        }
    }

    public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let location = locations.last else { return }

        // Filter out stale or inaccurate locations
        let age = Date().timeIntervalSince(location.timestamp)
        if age > 10 {
            logger.debug("Ignoring stale location (age: \(age)s)")
            return
        }

        if location.horizontalAccuracy < 0 {
            logger.debug("Ignoring invalid location (negative accuracy)")
            return
        }

        updateLocation(location)
        logger.debug("Location updated: \(location.coordinate.latitude), \(location.coordinate.longitude) (accuracy: \(location.horizontalAccuracy)m)")

        // Handle refresh request
        if isRefreshing {
            // Wait for a reasonably accurate location
            if location.horizontalAccuracy <= 20 {
                isRefreshing = false
                refreshContinuation?.resume(returning: location)
                refreshContinuation = nil

                // Restore balanced accuracy
                locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters

                if !isUpdating {
                    locationManager.stopUpdatingLocation()
                }
            }
        }

        // Reset timeout since we're getting updates
        if isUpdating {
            resetTimeoutTimer()
        }
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        logger.error("Location manager error: \(error.localizedDescription)")

        if isRefreshing {
            isRefreshing = false
            refreshContinuation?.resume(throwing: LocationError.locationUnavailable)
            refreshContinuation = nil

            locationManager.desiredAccuracy = kCLLocationAccuracyNearestTenMeters

            if !isUpdating {
                locationManager.stopUpdatingLocation()
            }
        }
    }
}

File: android/app/src/main/kotlin/com/marrakechguide/core/service/LocationService.kt
package com.marrakechguide.core.service

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.os.Looper
import android.util.Log
import androidx.core.content.ContextCompat
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import java.util.Timer
import java.util.TimerTask
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Permission status for location services.
 */
enum class PermissionStatus {
    /** Permission not yet requested */
    NOT_DETERMINED,
    /** Permission granted for foreground use */
    AUTHORIZED,
    /** Permission denied by user */
    DENIED,
    /** Permission restricted by system policy */
    RESTRICTED
}

/**
 * Errors that can occur during location operations.
 */
sealed class LocationError : Exception() {
    object PermissionDenied : LocationError() {
        private fun readResolve(): Any = PermissionDenied
        override val message = "Location permission was denied"
    }
    object LocationUnavailable : LocationError() {
        private fun readResolve(): Any = LocationUnavailable
        override val message = "Unable to determine your location"
    }
    object Timeout : LocationError() {
        private fun readResolve(): Any = Timeout
        override val message = "Location request timed out"
    }
    object ServicesDisabled : LocationError() {
        private fun readResolve(): Any = ServicesDisabled
        override val message = "Location services are disabled"
    }
}

/**
 * Interface for location services providing GPS coordinates.
 *
 * This service wraps FusedLocationProviderClient with battery-safe patterns:
 * - Request only foreground location (ACCESS_FINE_LOCATION)
 * - Start/stop updates based on screen visibility
 * - Automatic timeout to prevent battery drain
 * - Graceful handling of permission denial
 */
interface LocationService {
    /** Current user location, null if unknown or permission denied. */
    val currentLocation: StateFlow<Location?>

    /** Current permission status for location services. */
    val permissionStatus: StateFlow<PermissionStatus>

    /** Whether location services are authorized and available. */
    val isAuthorized: Boolean

    /**
     * Checks the current permission status without requesting.
     */
    fun checkPermission(): PermissionStatus

    /**
     * Starts receiving location updates.
     * Updates use balanced accuracy to preserve battery.
     * Call stopUpdates() when compass screen is no longer visible.
     */
    fun startUpdates()

    /**
     * Stops receiving location updates.
     * Always call this when the compass screen closes.
     */
    fun stopUpdates()

    /**
     * Requests a single high-accuracy location refresh.
     * Use sparingly as this is more battery-intensive.
     *
     * @return The refreshed location.
     * @throws LocationError if location cannot be obtained.
     */
    suspend fun refreshLocation(): Location
}

/**
 * Implementation of LocationService using FusedLocationProviderClient.
 *
 * This implementation:
 * - Uses PRIORITY_BALANCED_POWER_ACCURACY by default
 * - Escalates to HIGH_ACCURACY only for explicit refresh
 * - Includes automatic timeout after 10 minutes
 * - Throttles updates to reduce battery usage
 */
@Singleton
class LocationServiceImpl @Inject constructor(
    @ApplicationContext private val context: Context
) : LocationService {

    companion object {
        private const val TAG = "LocationService"
        private const val TIMEOUT_DURATION_MS = 600_000L // 10 minutes
        private const val UPDATE_INTERVAL_MS = 5_000L // 5 seconds
        private const val FASTEST_INTERVAL_MS = 2_000L // 2 seconds minimum
        private const val REFRESH_TIMEOUT_MS = 15_000L // 15 seconds for refresh
    }

    private val fusedClient: FusedLocationProviderClient =
        LocationServices.getFusedLocationProviderClient(context)

    private val _currentLocation = MutableStateFlow<Location?>(null)
    override val currentLocation: StateFlow<Location?> = _currentLocation.asStateFlow()

    private val _permissionStatus = MutableStateFlow(PermissionStatus.NOT_DETERMINED)
    override val permissionStatus: StateFlow<PermissionStatus> = _permissionStatus.asStateFlow()

    override val isAuthorized: Boolean
        get() = checkPermission() == PermissionStatus.AUTHORIZED

    private var isUpdating = false
    private var timeoutTimer: Timer? = null

    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(result: LocationResult) {
            result.lastLocation?.let { location ->
                // Filter out stale locations (older than 10 seconds)
                val age = System.currentTimeMillis() - location.time
                if (age > 10_000) {
                    Log.d(TAG, "Ignoring stale location (age: ${age}ms)")
                    return
                }

                // Filter out invalid locations
                if (location.accuracy <= 0) {
                    Log.d(TAG, "Ignoring invalid location (accuracy: ${location.accuracy})")
                    return
                }

                _currentLocation.value = location
                Log.d(TAG, "Location updated: ${location.latitude}, ${location.longitude} (accuracy: ${location.accuracy}m)")

                // Reset timeout since we're getting updates
                if (isUpdating) {
                    resetTimeoutTimer()
                }
            }
        }
    }

    override fun checkPermission(): PermissionStatus {
        val fineLocation = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_FINE_LOCATION
        )
        val coarseLocation = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.ACCESS_COARSE_LOCATION
        )

        val status = when {
            fineLocation == PackageManager.PERMISSION_GRANTED -> PermissionStatus.AUTHORIZED
            coarseLocation == PackageManager.PERMISSION_GRANTED -> PermissionStatus.AUTHORIZED
            else -> PermissionStatus.DENIED
        }

        _permissionStatus.value = status
        return status
    }

    @SuppressLint("MissingPermission")
    override fun startUpdates() {
        if (!isAuthorized) {
            Log.i(TAG, "Cannot start updates: not authorized")
            return
        }

        if (isUpdating) {
            Log.d(TAG, "Updates already active")
            return
        }

        isUpdating = true

        val locationRequest = LocationRequest.Builder(
            Priority.PRIORITY_BALANCED_POWER_ACCURACY,
            UPDATE_INTERVAL_MS
        ).apply {
            setMinUpdateIntervalMillis(FASTEST_INTERVAL_MS)
            setWaitForAccurateLocation(false)
        }.build()

        fusedClient.requestLocationUpdates(
            locationRequest,
            locationCallback,
            Looper.getMainLooper()
        )

        Log.i(TAG, "Started location updates")

        // Set timeout to prevent indefinite battery drain
        resetTimeoutTimer()
    }

    override fun stopUpdates() {
        if (!isUpdating) return

        isUpdating = false
        fusedClient.removeLocationUpdates(locationCallback)
        cancelTimeoutTimer()
        Log.i(TAG, "Stopped location updates")
    }

    @SuppressLint("MissingPermission")
    override suspend fun refreshLocation(): Location {
        if (!isAuthorized) {
            throw LocationError.PermissionDenied
        }

        // If we have a recent location, return it
        _currentLocation.value?.let { location ->
            val age = System.currentTimeMillis() - location.time
            if (age < 5_000) {
                return location
            }
        }

        // Request high-accuracy location
        return withTimeoutOrNull(REFRESH_TIMEOUT_MS) {
            suspendCancellableCoroutine { continuation ->
                val highAccuracyRequest = LocationRequest.Builder(
                    Priority.PRIORITY_HIGH_ACCURACY,
                    1_000L
                ).apply {
                    setMaxUpdates(1)
                    setWaitForAccurateLocation(true)
                }.build()

                val refreshCallback = object : LocationCallback() {
                    override fun onLocationResult(result: LocationResult) {
                        fusedClient.removeLocationUpdates(this)
                        val location = result.lastLocation
                        if (location != null) {
                            _currentLocation.value = location
                            if (continuation.isActive) {
                                continuation.resume(location)
                            }
                        } else {
                            if (continuation.isActive) {
                                continuation.resumeWithException(LocationError.LocationUnavailable)
                            }
                        }
                    }
                }

                @SuppressLint("MissingPermission")
                fun startRequest() {
                    fusedClient.requestLocationUpdates(
                        highAccuracyRequest,
                        refreshCallback,
                        Looper.getMainLooper()
                    )
                }
                startRequest()

                continuation.invokeOnCancellation {
                    fusedClient.removeLocationUpdates(refreshCallback)
                }
            }
        } ?: throw LocationError.Timeout
    }

    private fun resetTimeoutTimer() {
        cancelTimeoutTimer()

        timeoutTimer = Timer().apply {
            schedule(object : TimerTask() {
                override fun run() {
                    Log.w(TAG, "Location updates timed out after ${TIMEOUT_DURATION_MS}ms")
                    stopUpdates()
                }
            }, TIMEOUT_DURATION_MS)
        }
    }

    private fun cancelTimeoutTimer() {
        timeoutTimer?.cancel()
        timeoutTimer = null
    }
}
